### ISSUE TITLE
DynamicHooknameFound could check if the variable is a Tokens::$stringTokens and test

### ISSUE THREAD

**Issue by LC43** [WordPress/WordPress-Coding-Standards#2527]
## Bug Description

consider this:
```php
// setup
$filter = 'aaa';
$name   = 'bbb';
// the important part:
$hook   = "Saltus/{$filter}/filter-output";
do_action( $hook, $name, $filter );
```

its pretty obvious that the variable could be a string, or something that resembles a string.

why not check?

### Code sample

in vendor/wp-coding-standards/wpcs/WordPress/Sniffs/NamingConventions/PrefixAllGlobalsSniff.php:1037
```php
public function process_parameters( $stackPtr, $group_name, $matched_content, $parameters ) 
```

checks for a few things, but it just fails to follow up on the value, and just complains that `$hook` doesnt have the right prefix.

we could check if the variable has been defined, and its some sort of string. 
```php
$var_value = '';
$var_name = $this->tokens[ $first_non_empty ]['content'];
$assignmentPtr = $this->phpcsFile->findPrevious( T_VARIABLE, $stackPtr, null, false, $var_name );
if ($assignmentPtr) {
	$valuePtr = $this->phpcsFile->findNext( Tokens::$stringTokens, $assignmentPtr + 1);
	if ( isset( Tokens::$stringTokens[ $this->tokens[ $valuePtr ]['code'] ] ) ) {
		$first_non_empty_content = TextStrings::stripQuotes( $this->tokens[ $valuePtr ]['content'] );
		if ( $this->is_prefixed( $valuePtr, $first_non_empty_content ) === true ) {
			$var_value = $first_non_empty_content;
		}
	}
}

```

this works for these:

```php
$hook_with_vars   = "Saltus/{$name}/filter-output";
$hook_name_string = "Saltus/admin-cols/filter-output";
$hook_name_1quote = 'Saltus/admin-cols/filter-output';
```
## Minimal Code Snippet

```php
// setup
$filter = 'aaa';
$name   = 'bbb';
// the important part:
$hook   = "Saltus/{$filter}/filter-output";
do_action( $hook, $name, $filter );
```

## Error Code

```
Hook names invoked by a theme/plugin should start with the theme/plugin prefix. 
Found: "$hook_name_string".PHPCS( WordPress.NamingConventions.PrefixAllGlobals.DynamicHooknameFound )

```

## Custom Ruleset
any

## Expected behavior
Expect to validate the value of the variable if its possible.

## Environment

| Question                 | Answer
| ------------------------ | -------
| Operating System | [any]
| PHP version |  8.4
| Standard | [ wordpress ]
| Install type | [Composer (global/local) ]
| dealerdirect/phpcodesniffer-composer-installer | 1.0.0 
| phpcsstandards/phpcsextra                      | 1.2.1 
| phpcsstandards/phpcsutils                      | 1.0.12
| squizlabs/php_codesniffer                      | 3.12.0
| wp-coding-standards/wpcs                       | 3.1.0 
## Please confirm:

- [x] I have verified the issue still exists in the `develop` branch of WordPressCS.
- [x] I can create a pull request to implement this.

**Comment by jrfnl** [comment]
@LC43 I understand what you are saying and yes, in principle, the idea of this is a nice addition, however, in practice it will be neigh impossible to implement reliably, safe for very few cases (hard-coded variable defined once, in which case: why use a variable at all ?). And if it would only work for a few cases, is it worth the (large) effort to implement this ?

Keep in mind that PHPCS only has the context of the current file, variables can be defined and then added to or redefined, the variable may be a function parameter, class or global constant, function property, etc etc.

PHPCS is not a tool which can trace variables throughout a codebase, let alone trace the _likely_ _type_ of that variable.

**Comment by LC43** [comment]
thanks @jrfnl  for giving it a thought and encouraging words.

I don't have any statistics other than our code, but i imagine most cases will fall in these examples. 
Yes, i'm aware that there's a small scope of how previously we can search.

But at least for that majority of cases, this sniff will pass.

I'm guessing you are foreseeing a large effort but its a small check and if not finding the values, it can still default to the warning that was there before.
all the supporting values and functions are already in the code.
That conditional block already does the trick.

Care to explain where you see this turning more complicated?

**Comment by LC43** [comment]
im mostly trying to fix this:
`$hook   = "Saltus/{$filter}/filter-output";`

but it will also work on these:
`$hook_name_string = "Saltus/admin-cols/filter-output";`
`$hook_name_1quote = 'Saltus/admin-cols/filter-output';`

**Comment by jrfnl** [comment]
> I don't have any statistics other than our code, but i imagine most cases will fall in these examples.

That is not my experience.

> I'm guessing you are foreseeing a large effort but its a small check and if not finding the values, it can still default to the warning that was there before.

I listed a number of examples above. If any variable tracing is implemented, it should be done properly and that will involve a lot of checks and token walking, which will also severely impact performance of the sniff.

And it still wouldn't prevent the notice in most cases.

Consider the following examples which are all variations of your original sample:

```php
// This is doable.
$filter = 'aaa';
$name   = 'bbb';
$hook   = "Saltus/{$filter}/filter-output"; // Okay, we can see this is a string and can check the prefix part.
do_action( $hook, $name, $filter );

// This would already become more complicated:
$hook   = 'Saltus/' . $filter .'/filter-output'; // Now we'd have to deal with interpreting concatenation.
do_action( $hook, $name, $filter );

// And even more complicated (from a sniff point of view) 1:
$hook   = 'Saltus/';
$hook  .= $filter;
$hook  .= '/filter-output';
do_action( $hook, $name, $filter );

// And even more complicated (from a sniff point of view) 2:
if ($foo) {
    $hook = "Saltus/{$filter}/filter-output";
} else {
    $hook = $other_hook_name;
}
do_action( $hook, $name, $filter );

// And even more complicated (from a sniff point of view) 3:
$prefix = 'Saltus';
...other code....

$hook_parts = [
    $prefix,
    $filter,
    'filter-output',
];
$hook = implode( '/', $hook_parts);
do_action( $hook, $name, $filter );
```

Now also consider that some files are 1000+ lines long with code like the samples in the global namespace. Finding the variable may take walking to the very start of the file, only to conclude that the variable was not defined in this file.

Now, you may say that most of these things don't happen in the global namespace - and I wish you were right, but I've looked at too many crappy plugins to expect that to be true.
Even so, let's just say, for arguments sake, that nobody uses the global namespace anymore. Now consider the following:

```php
function my_action($prefix, $filter, $name) {
    $hook   = "$prefix/{$filter}/filter-output"; // We're out of luck.
    do_action( $hook, $name, $filter );
}

function my_action() {
    $hook   = CONSTANT_CONTAINING_PREFIX."/{$filter}/filter-output"; // We're out of luck.
    do_action( $hook, $name, $filter );
}
```

Do you need me to continue ?

**Comment by LC43** [comment]
fair! thanks for the enlightening post 🙏

**Comment by jrfnl** [comment]
@LC43 Sorry to be discouraging, but writing good sniffs is a little bit more complicated than most people think. Code can be written in so many different ways and for something like variable tracing, you really need to take all those ways into account, otherwise it will be a continuous game of whack-a-mole with bugs being reported left right and center.

**Comment by LC43** [comment]
not at all discouraging, i did ask for more details! i had fun looking into the code and learned something from you.
 thank you for taking time to expand on the use cases!
